Configurable query/prompt syntax

michaellavelle opened the pull request:
* Adding the ability to conifugure GPT with optional parameters (input_prefix, input_suffix, output_prefix, output_suffix, append_input_suffix_and_output_prefix_to_query,  stop) to allow the structure of the "input/output" questions to be changed - for example to use "Q/A" instead of input/output.   All optional parameters added for both GPT and Example default to their previous defaulted values.

* Adding a new example script ( run_latex_q_and_a_app.py ) to showcase the following configuration:
- Change of "input: " input_prefix to "Q: "
- Change of "output: " output_prefix to "A: "
- Change of stop from "'\ninput:" to "\nQ:"
- Appending "\nA: " ( input_suffix + output_prefix) to existing prompt query.

This allows the following type of prompts to be generated:

Q: Two plus two equals four
A: 2 + 2 = 4

Q: The integral from zero to infinity
A: \int_0^{\infty}

Q: The gradient of x squared plus two times x with respect to x
A: \nabla_x x^2 + 2x

Q: The log of two times x
A: \log{2x}

Q: x squared plus y squared plus equals z squared
A: x^2 + y^2 = z^2

Q: The sum from zero to twelve of i squared
A: \sum_{i=0}^{12} i^2
        
Q: E equals m times c squared
A: E = mc^2  

Q: H naught of t
A: H_0(t)
    
Q: f of n equals 1 over (b-a) if n is 0 otherwise 5
A: f(n) = \begin{cases} 1/(b-a) &\mbox{if } n \equiv 0 \\ # 5 \end{cases}
    
Q: x squared plus 2 times x
A:

shreyashankar commented on api/gpt.py:
```diff
@@ -10,9 +10,13 @@ def set_openai_key(key):
     """Stores an input, output pair and formats it to prime the model."""
 
-    def __init__(self, inp, out):
+    def __init__(self, inp, out, input_prefix = "input: ", input_suffix = "\n", output_prefix = "output: ", output_suffix = "\n"):
```
To minimize redundancy, I think it might be better to declare the prefixes and suffixes in the `GPT` class, and remove the `format` method from this class entirely. See my comment in the `GPT` class for an idea of how to format the example in there.

michaellavelle replied:
I've removed the formate function entirely, and all references to the prefix from Example.  I've added a new format_example function to GTP as suggested which is now called by the add_example function.

shreyashankar replied:
Awesome!

shreyashankar commented on api/gpt.py:
```diff
@@ -33,11 +38,23 @@ class GPT:
+        self.append_input_suffix_and_output_prefix_to_query = append_input_suffix_and_output_prefix_to_query
+        self.stop = stop
 
     def add_example(self, ex):
```
In this class, you could include a method such as:

```
def format_example(self, ex):
        """Formats the input, output pair."""
        return self.input_prefix + ex.get_input() + self.input_suffix + self.output_prefix + ex.get_output() + self.output_suffix
```

Then the `add_example` method could be:

```
def add_example(self, ex):
        """Adds an example to the object. Example must be an instance
        of the Example class."""
        assert isinstance(ex, Example), "Please create an Example object."
        self.examples.append(self.format_example(ex))
```

michaellavelle replied:
format_example method added which is now called by add_example as suggested

shreyashankar replied:
👍

shreyashankar commented on api/gpt.py:
```diff
@@ -74,7 +95,7 @@ def submit_request(self, prompt):
                                             n=1,
                                             stream=False,
-                                            stop="\ninput:")
+                                            stop=self.stop)
```
Nice idea :)

michaellavelle replied:
Thanks :-)

shreyashankar commented on examples/run_latex_q_and_a_app.py:
```diff
@@ -0,0 +1,45 @@
+          append_input_suffix_and_output_prefix_to_query = True,
+          stop = "\nQ:")
+
+gpt.add_example(Example('Two plus two equals four', '2 + 2 = 4', question_prefix, question_suffix, answer_prefix, answer_suffix))
```
If you remove the prefixes and suffixes from the `Example` class, then you could simplify to:

```
gpt.add_example(Example('Two plus two equals four', '2 + 2 = 4'))
```

michaellavelle replied:
Much cleaner - thanks!

shreyashankar commented on api/gpt.py:
```diff
@@ -33,11 +38,23 @@ class GPT:
+        self.input_suffix = input_suffix
+        self.output_prefix = output_prefix
+        self.output_suffix = output_suffix
+        self.append_input_suffix_and_output_prefix_to_query = append_input_suffix_and_output_prefix_to_query
```
Can you document or explain what this parameter, `append_input_suffix_and_output_prefix_to_query`,  could be used for? I understand why you may want the `output_prefix` to be optional in the priming text, but to format the priming properly, I think you may always need the `input_suffix` in the priming. But I could be wrong -- would love to hear your thoughts here!

michaellavelle replied:
Totally agree - I think I'd got in a muddle with this trying to replicate existing offsets ( see comments below about the change to the value now being sent for stop,  and the change to the offset).    I've now changed parameter name to append_output_prefix_to_query,  which now your suggested changes have been made is used to do exactly that - to add the output prefix only.

shreyashankar replied:
Sounds good!

michaellavelle commented on api/gpt.py:
```diff
@@ -33,11 +38,23 @@ class GPT:
+                 output_prefix = "output: ",
+                 output_suffix = "\n",
+                 append_input_suffix_and_output_prefix_to_query = False,
+                 stop = "\ninput:"):
```
Now removed stop parameter entirely and set self.stop as suggested.

shreyashankar replied:
Great!

michaellavelle commented on api/gpt.py:
```diff
@@ -63,7 +80,11 @@ def get_max_tokens(self):
     def craft_query(self, prompt):
         """Creates the query for the API request."""
-        return self.get_prime_text() + "input: " + prompt + "\n"
+        q = self.get_prime_text() + self.input_prefix + prompt
```
Now setting value of q as suggested

shreyashankar replied:
Great, thanks!

michaellavelle commented on api/gpt.py:
```diff
@@ -24,7 +28,8 @@ def get_output(self):
     def format(self):
         """Formats the input, output pair."""
-        return f"input: {self.input}\noutput: {self.output}\n"
+        return self.input_prefix + f"{self.input}" + self.input_suffix + self.output_prefix + f"{self.output}\n"
```
As per your later comments, this method has now been removed

shreyashankar replied:
Great, thanks!

michaellavelle commented:
Hi Shreya

Thanks so much for your comments and for the comprehensive review!  

I'd been concerned about possible regression, and had been trying to ensure that the existing functionality didn't break - I think the issues crept in because I'm using a mock of the GPT service to test against as I don't have access to the real thing yet.   Turns out there was a bug in my mocking server primarily to do with new line characters.   So my regression tests were testing against the wrong thing! 

I think I've fixed the issue with the mock server now (  https://github.com/ml4j/gpt-3-mock-api ).  Assuming this was the correct fix,  the existing scripts and the new script I've added for q_and_a both work for me - hope this issue is now resolved ?

Once I'd resolved the regression issue, your suggestions were just the thing I needed to do - I've now made all the changes and the code is much cleaner - many thanks!

Just a couple of points I noticed when testing.  The changes suggested to the stop parameter value seem to work - but the submitted request now contains the stop keyword without the newline prefix ( ie.  before the stop was "\ninput:" - where now it is just "input:"
I also noticed that for the original latex script value of offset was hardcoded to "7" but now I believe the calculated value is 8.    Everything seems to work - but just wanted to raise these changes in implementation in case they were problematic.   Let me know if so - I can have a go at reworking so the previous values are retained.

michaellavelle commented:
Also,  I made one additional commit for the following:

* Removing newline delimiter in join of examples (and the appended newline) , as each example already contains prefix and suffix whitespace information.  
* Changing default output_suffix to "\n\n" which adjusts for the join delimiter change and removal of appended newline.
* This allows configurations which may not separate inputs and outputs by lines ( eg. for blocks of continuous text)

I had been trying to integrate some of my existing templates with the project, and was getting stuck because they had output_suffixes of multiple newlines, eg.

https://github.com/ml4j/gtp-3-prompt-templates/blob/master/question-answer/default/templates/question_answer_template_1.json
https://github.com/ml4j/gtp-3-prompt-templates/blob/master/question-answer/code-explanation/templates/code_explanation_template_1.json 

This change allows these templates to be used, and also allows me to configure a setup with no newlines between one question/answer block and the next

michaellavelle commented:
Thanks Shreya.   I've replaced the latex_q_and_a example with a general knowedge q and a.   I hope this works - I don't have access to the api so I've been using AIDungeon - hard to know if the behaviour is the same - especially as there is no possiblitly of sending a stop word.

michaellavelle commented:
Hi Shreya.  I had to do a bit a refactoring after running pylint - primarily due to limitations on number of contructor parameters.  Two new simple classes have been introduced as a result - TextFormat and QueryFormat.   I initially tried to merge them into a single class but ran into the parameter limit again.   Also the separation allows TextFormat to be reused in multiple places - in building up the examples, and in constructing the prompt.

I've run pylint on the python files in the app package - no errors at all now for GPT,  and no regression of issues for the other files.   There seem to be some output such as this:

************* Module demo_web_app
demo_web_app.py:8:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
demo_web_app.py:9:0: E0402: Attempted relative import beyond top-level package (relative-beyond-top-level)
demo_web_app.py:4:0: W0611: Unused import openai (unused-import)

But I'm seeing these issues on the master branch too - not sure if they are false positives

shreyashankar commented:
Oh wow, I didn't realize. I should configure the linter to ignore the non-spacing errors. I don't think the `TextFormat` and `QueryFormat` classes are necessary -- what you had before is just fine. I will update your branch and merge. Thanks so much!
